### PR TITLE
Current development branch

### PR DESCRIPTION
**Pull request by mrossinek**
Consolidates the work on:

- the formal syntax specification (#29)
- the breaking changes and `paragraph` parsing refactoring (#30)
- the inline link targets (#25)

### CODE REVIEW

File: docs/NFF-0.1-spec.norg [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+    which are outlined in the later sections of this document.
+
+    A *word* is considered to be any combination of {> character}s which are neither {# whitespace},
+    {# punctuation} nor other {# special character}s.
+
+*** Paragraph Segments
+    {# Words} are first combined into {# paragraph segments}. A segment gets terminated by any of
+    the following:
+    - A {> line ending}
+    - Any of the {# attached modifiers} +NOTE(vhyrro): Don't paragraph segments contain attached mods?+
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:110]: Yes, you are right. Same with linkables actually.. I guess that means `paragraph_segment` is actually a line, right? :sweat_smile:

File: docs/NFF-0.1-spec.norg [hunk 2]
```diff
@@ -0,0 +1,778 @@
...
+
+** Attached Modifiers
+   We can now discuss attached modifiers (which originally gave rise to the name of
+   {# detached modifiers} as their natural counter-parts). Here are the general rules for attached
+   modifiers:
+   - An opening modifier may only be preceded by {# whitespace} or {# punctuation}
+   - An opening modifier may _NOT_ be succeeded by {# whitespace}
+   - A closing modifier may _NOT_ be preceded by {# whitespace}
+   - A closing modifier may only be succeeded by {# whitespace} or {# punctuation}
+   - {# Attached modifiers} can only span at maximum a single {# paragraphs}[paragraph].
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:473]: I don't fully get this statement? Didn't we say we don't want any new lines supported inside attached modifiers? If not, then what does `at maximum a single paragraph` mean? Do you mean that new lines are okay, but `paragraph_break`s are not? If yes, this should be more explicit.

File: docs/NFF-0.1-spec.norg [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+**** Custom Detached Modifiers
+     Apart from linking to the detached modifiers outlined above, you can also link to a set of
+     custom modifiers specifically designed for links. These are the `#` (magic) and `@` (file)
+     linkable locations.
+
+***** The File Linkable (`@`)
+      Sometimes you may want to link to an external file that isn't an _Norg_ file.
+      In that case you may use `{@ /path/to/my/file.txt}` (notice the mandatory space after the `@`
+      char, just like with the {**** detached modifier}s). Paths are relative to the _Norg_
+      file that contains the link.
```
[the remark this answers is not in the document]
Comment (vhyrro) [reply docs/NFF-0.1-spec.norg]: No I'm pretty sure we wanted this syntax from the very beginning lol

Even the treesitter tests had `{@ this.txt}` as a valid and passing case.
It wouldn't make sense to use `{@myfile.txt}`, since that would make it the only char that is unnecessarily inconsistent with everything else (even the magic char, `#`, requires whitespace).

File: docs/NFF-0.1-spec.norg (hunk 2 above)
Comment (vhyrro) [reply docs/NFF-0.1-spec.norg]: When we were discussing ranged attached modifiers I said that I'd not be a fan of allowing paragraph breaks within them, but I did say that allowing regular newlines within attached modifiers was okay.

By "spans at maximum a single paragraph", I meant:
```
*this

is not allowed*
```
However:
```
*this
is fine*
```

However you're right in saying that it may be a little confusing to someone reading it without a proper example. I'll change that.

File: docs/NFF-0.1-spec.norg [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+
+     ^ Blank lines are irrelevant in _Norg_ because standalone whitespace has no semantic relevance.
+
+     [Unicode Zs general category]{https://www.fileformat.info/info/unicode/category/Zs/list.htm}
+
+**** Punctuation
+     A {> character} is considered to be *punctuation* if it is any of the following:
+     - A standard ASCII punctuation character: |`!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~`|
+
+       +TODO: check whether we want to support every punctuation mark from the `en_US.utf8` locale+
```
[the remark this answers is not in the document]
Comment (vhyrro) [reply docs/NFF-0.1-spec.norg:80]: This may be a bit of a tricky one. I've resorted to looking at the commonmark spec to see how it's done over there:

> An ASCII punctuation character is !, ", #, $, %, &, ', (, ), *, +, ,, -, ., / (U+0021–2F), :, ;, <, =, >, ?, @ (U+003A–0040), [, \, ], ^, _, ` (U+005B–0060), {, |, }, or ~ (U+007B–007E).
>
> A Unicode punctuation character is an ASCII punctuation character or anything in the general Unicode categories Pc, Pd, Pe, Pf, Pi, Po, or Ps.

We may perhaps be interested in supporting a similar (if not the same) range of punctuation chars - these will probably be automatically covered via `std::iswpunct`, which will use the current locale.
The problem I can see arising though is minor inconsistencies or `std::iswpunct` supporting more than what we'd expect.

A possible solution to this could be to enforce `en_US.utf8` as the only natively supported locale - this can be done through `std::setlocale(LC_ALL, "en_US.utf8");` - but then we could run into issues with people using "other" punctuation chars like you said.

We may want to discuss this on discord as it's a pretty difficult thing to get right. The org syntax draft does not specify what punctuation is :(

File: docs/NFF-0.1-spec.norg (hunk 1 above)
Comment (vhyrro) [reply docs/NFF-0.1-spec.norg:104]: Yeah essentially a `paragraph_segment` is just a line of text :P

I can rephrase this to be a little more accurate.

File: docs/NFF-0.1-spec.norg [hunk 5]
```diff
@@ -0,0 +1,875 @@
...
+    @end
+
+*** Insertions
+    @code norg
+    = Insertion arg1 arg2 ...
+    @end
+    Depending on the kind of insertion, a different number of arguments can be set/required. These
+    arguments follow the same rules as those of {# Tags}.
+    +TODO: not sure I really like this. Tags are separate from detached modifiers, we shouldn't
+    draw parallels here+
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:392]: I disagree. Simply because having different kinds of rule sets simple to avoid this parallel would be confusing.

File: docs/NFF-0.1-spec.norg [hunk 6]
```diff
@@ -0,0 +1,875 @@
...
+    -- Thus, the start node of this extension can be seen as a {# trailing modifier}
+    - The contents follow
+    - The extension ends when:
+    -- Another {# nestable detached modifier} occurs
+    -- Or when the indentation is manually closed with one of the {# delimiting modifiers}
+     +TODO(vhyrro): discuss what truly "automatically" ends the indent segment. I propose that the only way
+     to auto-close an indent segment is to use a structural detached modifier or a modifier of the same
+     type as the parent of the indent segment; i.e. if you attach an indent segment to a level 1
+     unordered list item then creating another level 1 unordered list item auto-closes the indent
+     segment. However, creating a level 1 ordered list segment inside would be okay.+
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:439]: I agree with this :+1:

File: docs/NFF-0.1-spec.norg [hunk 7]
```diff
@@ -0,0 +1,875 @@
...
+*** Indent Segment Extension
+    Since {# nestable detached modifiers} can only contain a {# paragraph} this can cause severe
+    limitations because insertion of (for example) code blocks is not possible. To alleviate this
+    deficiency, the {# indent segment extension} exists, which increases the indentation of the
+    tree by one level allowing nesting of more complex elements to occur.
+    The syntax is the following:
+    - The indent segment extension starts with a single `\` char
+    - This must be _immediately_ followed by a {# line ending}
+    -- Thus, the start node of this extension can be seen as a {# trailing modifier}
+    - The contents follow
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:431]: Again: we need to be explicit

File: docs/NFF-0.1-spec.norg [hunk 8]
```diff
@@ -0,0 +1,875 @@
...
+      @code lua
+      print("Even code blocks are allowed!")
+      @end
+
+      ^ Range-able detached modifiers can be inserted, too.
+
+      ---
+    - Another nestable detached modifier ends the indent segment
+    - However, in this example I already closed the indent segment manually with `---`
+    @end
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:456]: This syntax currently does not work because the block ends prematurely when encountering the `@end` of the internal `@code lua` block. We need to switch to an `#example` tag instead (this holds practically for all `@code norg` blocks in this document).

File: docs/NFF-0.1-spec.norg [hunk 9]
```diff
@@ -0,0 +1,875 @@
...
+    and/or `-`, `_` and `.`.
+  - Tags can have their names delimited by a `.` in order to create a "hierarchy", e.g.
+    `document.meta`.
+  - \
+    After a whitespace character any number of parameters on the same line may follow:
+    @code norg
+    |tag-name.subtag parameter1 parameter2
+    @end
+
+    Parameters may consist of any character (apart from a {# line break}, of course).
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:478]: We need to discuss multi-word parameters here and the fact that this disallows the use of double quotes `"` in parameters

File: docs/NFF-0.1-spec.norg [hunk 10]
```diff
@@ -0,0 +1,875 @@
...
+    ---
+
+   _Norg_ provides several inbuilt tag names that are reserved, but their details are not explained
+   in this specification - this document strictly covers syntax. There is no restriction in regards
+   to the length of a tag name, nor are there any disallowed names that a parser should omit
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:483]: What about `@end` and `#end`? I feel like we should mention them.

Having said that I had another idea to further avoid issues with unexpected closing via those tags: could we support `end` and `END` as separate closing keywords which close tags based on whether or not the opening tag was lower- or upper-case? Just an idea, feel free to shoot it down.

File: docs/NFF-0.1-spec.norg [hunk 11]
```diff
@@ -0,0 +1,875 @@
...
+   They begin with the traditional tag declaration and are ended with an `end` statement.
+   They allow you to create your own completely custom syntax within a determined boundary.
+
+   The `end` statement has a simple ruleset:
+   - Must be at the start of a line, may be preceded by any whitespace (but nothing else)
+   - Must use the same prefix as its parent (in the case of regular ranged tags: `#`; in the case
+     of verbatim tags: `@`)
+   - Must not be succeeded by anything other than a {# line break}.
+
+*** Ranged Tags
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:504]:
```suggestion
*** Standard Ranged Tags
```
Just a suggestion to de-duplicate the heading

Comment (mrossinek) [reply docs/NFF-0.1-spec.norg:504]: I just saw that below you talk about "regular" rather than "standard" ranged tags. In the top of the document we talked about "standard detached mods". Shall we stick to one phrasing?

File: docs/NFF-0.1-spec.norg [hunk 12]
```diff
@@ -0,0 +1,875 @@
...
+  - \/italic\/: /italic/
+  - \_underline\_: _underline_
+  - \-strikethrough\-: -strikethrough-
+  - \!spoiler\!: !spoiler!
+  - \^superscript\^: ^superscript^  (cannot be nested into `subscript`)
+  - \,subscript\,: ,subscript,  (cannot be nested into `superscript`)
+  - \`inline code\`: `inline code` (disables any nested markup - verbatim)
+  - \+inline comment\+: +inline comment+
+  - \$inline math\$: $f(x) = y$  (verbatim)
+  - \=variable\=: =variable=  (verbatim)
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:577]: Needs a connection to insertions

File: docs/NFF-0.1-spec.norg [hunk 13]
```diff
@@ -0,0 +1,875 @@
...
+  +TODO: Examples+
+
+** Link Location
+   The link location is defined through curly braces (`{}`) and contains the physical location
+   that the user would like to link to. Inside of these curly braces you can find one of four
+   types of data:
+   - A {# file location}
+   - A URL (most commonly to an external resource)
+   - A detached modifier followed by the name of the linkable
+   - A custom detached modifier specifically made for links (`@`, `#`)
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:708]: Why is the first a link while the others are not?

File: docs/NFF-0.1-spec.norg [hunk 14]
```diff
@@ -0,0 +1,875 @@
...
+
+    `@` and URLs are not allowed:
+
+    @code norg
+    {:path:@ file} <- invalid
+    {:path:https://my-url} <- also invalid
+    @end
+
+*** URL
+    You can define a link to an external resource by simply putting in the URL:
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:748]: Should also discuss `file://` scheme and the fact that `xdg-open`/`open` (on Mac) is used for opening these files whereas neovim will be used for `@ external_file` link targets to explain the difference (I know this is not strictly syntax but the difference between the two should be pointed out somehow)

File: docs/NFF-0.1-spec.norg [hunk 15]
```diff
@@ -0,0 +1,875 @@
...
+   @end
+
+** Anchors
+   _Norg_ also has a concept called {# anchors}. These allow you to place a standalone
+   {# link description} inside of text (referred to as an <anchor declaration>).
+   The target which this anchor links to can then be _defined_ at another place in the document
+   with an <anchor definition> which is an initial {# link description} followed by a
+   {# link location}. This is especially useful when you want to link to the same target very
+   often, like for example a specific website.
+
```
Comment (mrossinek) [inline docs/NFF-0.1-spec.norg:826]: We need to talk about custom link descriptions in anchor definitions: `[anchor][my custom text]`

File: grammar.js [hunk 16]
```diff
@@ -1787,18 +749,14 @@ module.exports = grammar({
             )
         ),
 
-        // TODO: Comment scanner code
         ranged_tag_content: $ =>
```
Comment (mrossinek) [inline grammar.js:752]: The `ranged_tag_content` and `ranged_verbatim_tag_content` still need some work. Especially they need to be handled properly by the injections in https://github.com/nvim-neorg/neorg/pull/337 where a trivial change of `@table` to `#table` does not result in the desired effect as shown below:
```norg
#table
*bold heading* | /italic heading/
some value | _some underlined value_
#end
```
leads to an AST when the injection is updated:
```
ranged_tag [0, 0] - [3, 4]
  name: tag_name [0, 1] - [0, 6]
    tag_name_element [0, 1] - [0, 6]
  content: ranged_tag_content [1, 0] - [3, 0]
    row [1, 0] - [2, 0]
      cell [1, 0] - [1, 14]
        paragraph [1, 0] - [1, 14]
          paragraph [1, 0] - [1, 14]
            paragraph_segment [1, 0] - [1, 14]
              bold [1, 0] - [1, 14]
      cell [1, 17] - [1, 33]
        paragraph [1, 17] - [1, 33]
          paragraph [1, 17] - [1, 33]
            paragraph_segment [1, 17] - [1, 33]
              italic [1, 17] - [1, 33]
    row [2, 0] - [3, 0]
      cell [2, 0] - [2, 10]
        paragraph [2, 0] - [2, 10]
          paragraph [2, 0] - [2, 10]
            paragraph_segment [2, 0] - [2, 10]
      cell [2, 13] - [2, 36]
        paragraph [2, 13] - [2, 36]
          paragraph [2, 13] - [2, 36]
            paragraph_segment [2, 13] - [2, 36]
              underline [2, 13] - [2, 36]
    paragraph_segment [1, 0] - [1, 33]
      bold [1, 0] - [1, 14]
      italic [1, 17] - [1, 33]
    paragraph_segment [2, 0] - [2, 36]
      underline [2, 13] - [2, 36]
  ranged_tag_end [3, 0] - [3, 4]
```
We need to either: remove the nested injection of `norg` inside of `norg_table` but this will still treat the entire table as a `paragraph` from the `norg` parser perspective..
Or figure out a better solution..

File: docs/NFF-0.1-spec.norg [hunk 17]
```diff
@@ -0,0 +1,875 @@
...
+   {# punctuation} nor other {# special character}s.
+
+** Paragraph Segments
+   {# Words} are first combined into {# paragraph segments}. A segment may contain any inline
+   element:
+   - Any of the {# attached modifiers}
+   - Any of the {# linkables}
+   
+   A {# line ending} terminates the paragraph segment. This means that a paragraph segment is
+   essentially just a line of text:
```
[the remark this answers is not in the document]
Comment (vhyrro) [reply docs/NFF-0.1-spec.norg:116]: Although this used to be the case, this isn't entirely the case anymore, note that
```
* This *isn't
  bold text*
```

File: docs/NFF-0.1-spec.norg [hunk 18]
```diff
@@ -0,0 +1,875 @@
...
+
+*** Unordered Lists
+    @code norg
+    - Unordered list level 1
+    -- Unordered list level 2
+    --- Unordered list level 3
+    ---- Unordered list level 4
+    ----- Unordered list level 5
+    ------ Unordered list level 6
+    ------- Unordered list level 7 (falls back to level 6 in the tree-sitter parser)
```
[the remark this answers is not in the document]
Comment (vhyrro) [reply docs/NFF-0.1-spec.norg:181]: Yeah, I decided to do this just to make it very explicit that you can still put a single paragraph segment after the nestable detached modifier.
